Some models can't be converted to ModelRepr

A model like `Model.Vector(Reified[Model].model)` can't be converted to `ModelRepr`. When `ModelRepr.fromModel` is called, it throws an exception: `java.lang.AssertionError: tail of CCons is not a sum`.

--- a/circe/src/test/scala/dev/tauri/seals/circe/ModelCodecSpec.scala
+++ b/circe/src/test/scala/dev/tauri/seals/circe/ModelCodecSpec.scala
@@ -1,5 +1,7 @@
 /*
  * Copyright 2016-2020 Daniel Urban and contributors listed in AUTHORS
+ * Copyright 2020 Nokia
+ * SPDX-License-Identifier: Apache-2.0
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -93,9 +95,18 @@ class ModelCodecSpec extends BaseJsonSpec {
     )
   }
 
-  def ref(id: Int): Json = {
+  def sumRef(id: Int): Json =
+    mkRef(id, "SumRef")
+
+  def prodRef(id: Int): Json =
+    mkRef(id, "ProdRef")
+
+  def otherRef(id: Int): Json =
+    mkRef(id, "OtherRef")
+
+  def mkRef(id: Int, name: String): Json = {
     Json.obj(
-      "Ref" -> Json.obj(
+      name -> Json.obj(
         "id" -> Json.fromString(id.toString)
       )
     )
@@ -227,9 +238,9 @@ class ModelCodecSpec extends BaseJsonSpec {
         "HCons" -> Json.obj(
           "id" -> Json.fromString("0"),
           "label" -> Json.fromString("m"),
-          "refinement" -> jNone,
           "optional" -> jfalse,
-          "head" -> ref(2),
+          "refinement" -> jNone,
+          "head" -> sumRef(2),
           "tail" -> Json.obj(
             "HNil" -> Json.obj()
           )
@@ -307,26 +318,26 @@ class ModelCodecSpec extends BaseJsonSpec {
   "Encoding" - {
 
     "atoms" in {
-      a1.asJson should === (atom[Int])
-      a2.asJson should === (atom[String])
+      a1.asJson.spaces2 should === (atom[Int].spaces2)
+      a2.asJson.spaces2 should === (atom[String].spaces2)
     }
 
     "simple models" in {
-      m1.asJson should === (m1json)
-      m2.asJson should === (m2json)
+      m1.asJson.spaces2 should === (m1json.spaces2)
+      m2.asJson.spaces2 should === (m2json.spaces2)
     }
 
     "cyclic models" in {
-      cy1.asJson should === (cy1json)
-      cy2.asJson should === (cy2json)
+      cy1.asJson.spaces2 should === (cy1json.spaces2)
+      cy2.asJson.spaces2 should === (cy2json.spaces2)
     }
 
     "optional fields" in {
-      optMod.asJson should === (optJson)
+      optMod.asJson.spaces2 should === (optJson.spaces2)
     }
 
     "refined models" in {
-      m1r.asJson should === (m1rJson)
+      m1r.asJson.spaces2 should === (m1rJson.spaces2)
     }
   }
 
@@ -448,7 +459,7 @@ class ModelCodecSpec extends BaseJsonSpec {
             "optional" -> jfalse,
             "refinement" -> jNone,
             "head" -> Json.obj(
-              "Ref" -> Json.obj(
+              "ProdRef" -> Json.obj(
                 "id" -> Json.fromString("0")
               )
             ),
@@ -470,7 +481,7 @@ class ModelCodecSpec extends BaseJsonSpec {
               "optional" -> jfalse,
               "refinement" -> jNone,
               "head" -> Json.obj(
-                "Ref" -> Json.obj(
+                "ProdRef" -> Json.obj(
                   "id" -> Json.fromInt(56) // not string
                 )
               ),
@@ -491,7 +502,7 @@ class ModelCodecSpec extends BaseJsonSpec {
               "optional" -> jfalse,
               "refinement" -> jNone,
               "head" -> Json.obj(
-                "Ref" -> Json.obj(
+                "ProdRef" -> Json.obj(
                   "id" -> Json.fromString("abc") // NumberFormatException
                 )
               ),
@@ -512,7 +523,7 @@ class ModelCodecSpec extends BaseJsonSpec {
               "optional" -> jfalse,
               "refinement" -> jNone,
               "head" -> Json.obj(
-                "Ref" -> Json.obj(
+                "ProdRef" -> Json.obj(
                   "id" -> Json.fromString("999") // no target
                 )
               ),
@@ -533,7 +544,7 @@ class ModelCodecSpec extends BaseJsonSpec {
               "optional" -> jfalse,
               "refinement" -> jNone,
               "head" -> Json.obj(
-                "Ref" -> Json.obj(
+                "ProdRef" -> Json.obj(
                   "id" -> Json.fromString("-1") // no target
                 )
               ),
@@ -573,7 +584,7 @@ class ModelCodecSpec extends BaseJsonSpec {
               "optional" -> jfalse,
               "refinement" -> jNone,
               "head" -> Json.obj(
-                "Ref" -> Json.obj(
+                "ProdRef" -> Json.obj(
                   "id" -> Json.fromString("0") // forward ref
                 )
               ),

--- a/core/src/main/scala/dev/tauri/seals/core/model.scala
+++ b/core/src/main/scala/dev/tauri/seals/core/model.scala
@@ -1,5 +1,7 @@
 /*
  * Copyright 2016-2020 Daniel Urban and contributors listed in AUTHORS
+ * Copyright 2020 Nokia
+ * SPDX-License-Identifier: Apache-2.0
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -17,9 +19,11 @@
 package dev.tauri.seals
 package core
 
+import java.util.UUID
+
 import scala.annotation.tailrec
 import scala.util.hashing.MurmurHash3
-import java.util.UUID
+
 import cats.{ Eq, Show }
 import cats.data.State
 import cats.implicits._
@@ -146,7 +150,7 @@ sealed trait Model extends Serializable {
   private[this] type StMapP = State[MapP, Unit]
 
   @transient
-  private[this] lazy val cachedPathsAndIds = {
+  private[this] lazy val cachedPathsAndIds: Map[Model, (Model.Path, Int)] = {
     def compositePre(
       label: String,
       c: Model.Ctx,

--- a/core/src/main/scala/dev/tauri/seals/core/modelRepr.scala
+++ b/core/src/main/scala/dev/tauri/seals/core/modelRepr.scala
@@ -1,5 +1,7 @@
 /*
  * Copyright 2016-2020 Daniel Urban and contributors listed in AUTHORS
+ * Copyright 2020 Nokia
+ * SPDX-License-Identifier: Apache-2.0
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -30,7 +32,7 @@ private sealed trait ModelRepr {
   import ModelRepr._
 
   final def toModel: Either[String, Model] = {
-    val res = toModelSt.value.runA(Map.empty).value
+    val res = toModelSt.value.runA(HMap.empty).value
     // do a full traverse, to allow
     // dropping thunks and to ensure
     // that any bug in the decoding
@@ -48,12 +50,17 @@ private object ModelRepr extends ModelReprBase {
 
   private type Error = String
   private type DecSt[A] = EitherT[State[DecMap, ?], Error, A]
-  private type DecMap = Map[Int, Model]
+  type DecMap = HMap[DecMapRel]
+
+  final class DecMapRel[K, V]
+  final object DecMapRel {
+    implicit def ref[M <: Model]: DecMapRel[Ref[M], M] = new DecMapRel
+  }
 
   // TODO: try to simplify
-  sealed trait Composite[C <: Model, T <: Model, TR <: ModelRepr] { this: ModelRepr =>
+  sealed trait Composite[C <: T, T <: Model, TR <: ModelRepr] { this: ModelRepr =>
 
-    def id: Int
+    def ref: Ref[T]
     def head: ModelRepr
     def tail: TR
 
@@ -71,7 +78,7 @@ private object ModelRepr extends ModelReprBase {
           Eval.later(h.value._2.getOrElse(impossible(sh"accessing ${desc} parent of invalid `h`"))),
           Eval.later(t.value._2.getOrElse(impossible(sh"accessing ${desc} parent of invalid `t`")))
         )
-        lazy val newSt: DecMap = st + (id -> res)
+        lazy val newSt: DecMap = st + (ref -> (res : T))
         lazy val h = head.toModelSt.value.run(newSt)
         lazy val lh = Eval.later(h.value._2)
         lazy val t = decTail(tail).value.run(h.value._1)
@@ -95,7 +102,7 @@ private object ModelRepr extends ModelReprBase {
                   // either `h` or `t` failed, so we
                   // must remove the (invalid) inserted
                   // composite instance:
-                  State.modify[DecMap](_ - id).map(_ => res)
+                  State.modify[DecMap](_ - ref).map(_ => res)
                 },
                 _ => State.pure(res)
               )
@@ -127,6 +134,8 @@ private object ModelRepr extends ModelReprBase {
     head: ModelRepr,
     tail: ProdRepr
   ) extends ProdRepr with Composite[Model.HCons[_], Model.HList, ProdRepr] {
+
+    override val ref: Ref[Model.HList] = ProdRef(id)
 
     protected override val desc = "HCons"
 
@@ -162,6 +171,8 @@ private object ModelRepr extends ModelReprBase {
     tail: SumRepr
   ) extends SumRepr with Composite[Model.CCons, Model.Coproduct, SumRepr] {
 
+    override val ref: Ref[Model.Coproduct] = SumRef(id)
+
     protected override val desc = "CCons"
 
     protected override def build(h: Eval[Model], t: Eval[Model.Coproduct]): Model.CCons =
@@ -184,12 +195,25 @@ private object ModelRepr extends ModelReprBase {
       EitherT.fromEither(Right(Model.Atom(id, desc)))
   }
 
-  final case class Ref(id: Int) extends ModelRepr {
-    protected override def toModelSt: DecSt[Model] = {
+  // TODO: maybe optimize encoding, to on the wire all refs are a simple `Ref(id)`
+  sealed abstract class Ref[M <: Model](val id: Int) extends ModelRepr {
+    protected def toModelStImpl: DecSt[M] = {
       EitherT(State.get[DecMap].map { map =>
-        Either.fromOption(map.get(id), sh"invalid ID: $id")
+        Either.fromOption(map.get(this), sh"invalid ID: $id")
       })
     }
+  }
+
+  final case class ProdRef(override val id: Int) extends Ref[Model.HList](id) with ProdRepr {
+    override def toProdSt: DecSt[Model.HList] = toModelStImpl
+  }
+
+  final case class SumRef(override val id: Int) extends Ref[Model.Coproduct](id) with SumRepr {
+    override def toSumSt: DecSt[Model.Coproduct] = toModelStImpl
+  }
+
+  final case class OtherRef(override val id: Int) extends Ref[Model](id) {
+    protected final override def toModelSt = toModelStImpl
   }
 
   def fromModel(model: Model): ModelRepr = {
@@ -221,7 +245,11 @@ private object ModelRepr extends ModelReprBase {
         val id = map.get(c.m).getOrElse {
           impossible(sh"no ID found for cycle at ${c.p} (map is ${map})")
         }
-        Ref(id)
+        c.m match {
+          case _: Model.CCons | Model.CNil => SumRef(id)
+          case _: Model.HCons[_] | Model.HNil => ProdRef(id)
+          case _ => OtherRef(id)
+        }
       }
     )
   }

--- a/tests/src/test/scala/dev/tauri/seals/core/ModelReprSpec.scala
+++ b/tests/src/test/scala/dev/tauri/seals/core/ModelReprSpec.scala
@@ -1,0 +1,54 @@
+/*
+ * Copyright 2016-2020 Daniel Urban and contributors listed in AUTHORS
+ * Copyright 2020 Nokia
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.tauri.seals
+package core
+
+import cats.implicits._
+
+import org.scalatest.compatible.Assertion
+
+class ModelReprSpec extends tests.BaseSpec {
+
+  def roundtrip(m: Model): Assertion = {
+    val r = ModelRepr.fromModel(m)
+    val m2 = r.toModel.fold(err => fail(sh"cannot decode ModelRepr: ${err}"), m => m)
+    assert(Model.modelEquality.eqv(m2, m))
+    val r2 = ModelRepr.fromModel(m2)
+    assert(r2 === r)
+  }
+
+  "Cycles/Refs" in {
+    val modelModel = Reified[Model].model
+    val modelUnderTest = Model.Vector(
+      Model.HCons(
+        '_1,
+        Atomic[Int].atom,
+        Model.HCons(
+          '_2,
+          modelModel,
+          Model.HNil
+        )
+      )
+    )
+
+    roundtrip(modelModel)
+    roundtrip(Model.Vector(modelModel))
+    roundtrip(modelUnderTest)
+  }
+}

